src/workerd/api/filesystem.h

```diff
@@ -0,0 +1,333 @@
+  // they know through which path they are accessible through. Support for this
+  // is not critical so we likely won't implement this before we ship the feature
+  // but it is something we should consider doing in the future.
+  jsg::Promise<kj::Array<jsg::USVString>> resolve(
+      jsg::Lock& js, jsg::Ref<FileSystemHandle> possibleDescendant);
+
+  JSG_ITERATOR(
+      EntryIterator, entries, EntryIteratorType, IteratorState, iteratorNext<EntryIteratorType>);
+  JSG_ITERATOR(KeyIterator, keys, KeyIteratorType, IteratorState, iteratorNext<KeyIteratorType>);
+  JSG_ITERATOR(
+      ValueIterator, values, ValueIteratorType, IteratorState, iteratorNext<ValueIteratorType>);
+
+  void forEach(jsg::Lock& js,
+      jsg::Function<void(
+          jsg::USVString, jsg::Ref<FileSystemHandle>, jsg::Ref<FileSystemDirectoryHandle>)>
+          callback,
+      jsg::Optional<jsg::Value> thisArg);
+
+  JSG_RESOURCE_TYPE(FileSystemDirectoryHandle) {
+    JSG_INHERIT(FileSystemHandle);
+    JSG_METHOD(getFileHandle);
+    JSG_METHOD(getDirectoryHandle);
+    JSG_METHOD(removeEntry);
+    JSG_METHOD(resolve);
+    JSG_METHOD(entries);
+    JSG_METHOD(keys);
+    JSG_METHOD(values);
+    JSG_ITERABLE(entries);
+    JSG_METHOD(forEach);
+  }
+
+ private:
+  kj::Rc<workerd::Directory> inner;
+
+  template <typename Type>
+  static kj::Maybe<Type> iteratorNext(jsg::Lock& js, IteratorState& state) {
+    if (state.index >= state.entries.size()) {
+      return kj::none;
+    }
+
+    auto& entry = state.entries[state.index++];
+
+    if constexpr (kj::isSameType<Type, EntryIteratorType>()) {
+      return EntryType(js.accountedUSVString(entry->getName(js)), entry.addRef());
+    } else if constexpr (kj::isSameType<Type, KeyIteratorType>()) {
+      return js.accountedUSVString(entry->getName(js));
+    } else if constexpr (kj::isSameType<Type, ValueIteratorType>()) {
+      return entry.addRef();
+    } else {
+      static_assert(false, "invalid iterator type");
+    }
+  }
+};
+
+class FileSystemWritableFileStream: public WritableStream {
+  // TODO(node-fs): The spec defines FileSystemWritableFileStream objects such
+  // that any changes to the file are not actually committed until the stream
+  // is closed. We're not yet implementing this. All writes modify the source
+  // file in memory and would be immediately visible to other handles pointing
+  // to the same file. This should be updated before this feature is unflagged
+  // and made stable.
+ public:
+  struct State: public kj::Refcounted {
+    kj::Maybe<kj::Rc<workerd::File>> file;
+    size_t position = 0;
+    State(kj::Rc<workerd::File> file): file(kj::mv(file)) {}
+  };
+
+  FileSystemWritableFileStream(
+      kj::Own<WritableStreamController> controller, kj::Rc<State> sharedState);
+
+  static jsg::Ref<FileSystemWritableFileStream> constructor() = delete;
+
+  struct WriteParams {
+    kj::String type;  // one of: write, seek, truncate
```

Comment:
General nit:
Why not use JsStrings everywhere?
Non blocking of course

Reply:
those still end up needing to be converted to `kj::String` for comparison operations like what we see in `FileSystemWritableFileStream::write` (e.g. `if (params.type == "write"_kj) {` unless we created `JsStrings` to replace the `"write"_kj` part. I'd rather keep this as familiar as possible to our existing convention. If we want to convert more of the `kj::String` uses to `JsString` en masse then we can do so in a separate PR.